### PR TITLE
remove use of deprecated size function

### PR DESCRIPTION
**Pull request by correaa**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Remove use of now deprecated function in Multi

## What type(s) of changes does this code introduce?

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

- Yes This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/AFQMC/HamiltonianOperations/THCOps.hpp [hunk 1]
```diff
@@ -133,43 +133,43 @@ class THCOps
         vn0(std::move(v0_)),
         E0(e0_)
   {
-    gnmu    = Luv.size(1);
-    grotnmu = rotMuv.size(1);
+    gnmu    = std::get<1>(Luv.sizes());
```
Comment (PDoakORNL) [inline src/AFQMC/HamiltonianOperations/THCOps.hpp:136]: well its ugly, but I imagine you had your reasons to drop the `size_t size(int)` interface.
I assume there is no performance impact or an improvement.  I guess you also get compile time bounds checking.

**Review by PDoakORNL — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

File: src/AFQMC/HamiltonianOperations/THCOps.hpp (hunk 1 above)
Comment (correaa) [reply src/AFQMC/HamiltonianOperations/THCOps.hpp:136]: yes, this is a current defect in C++, it is ugly to access aggregates by index, but the code advantages are there: 
The compile-time access to the index allows the compiler to generate the internal sizes as necessary.
Also, compile-time access allows to check for bound errors (in the number of dimensions) at compile-time.

**Comment by correaa** [comment]
ok to merge for now.

**Comment by ye-luo** [comment]
Test this please

File: src/AFQMC/Matrix/csr_matrix_construct.hpp [hunk 2]
```diff
@@ -70,11 +70,11 @@ CSR construct_csr_matrix_single_input(MultiArray2D&& M, double cutoff, char TA,
     }
     else
     {
-      nr = M.size(1);
-      nc = M.size(0);
+      nr = std::get<1>(M.sizes());
+      nc = std::get<1>(M.sizes());
```
Comment (ye-luo) [inline src/AFQMC/Matrix/csr_matrix_construct.hpp:74]: suspicious change.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — DISMISSED** [review]

File: src/AFQMC/Matrix/csr_matrix_construct.hpp (hunk 2 above)
Comment (correaa) [reply src/AFQMC/Matrix/csr_matrix_construct.hpp]: yes, thank you for (train)spotting this.

**Comment by correaa** [comment]
It seems that some intel test is failing,
https://github.com/QMCPACK/qmcpack/runs/8047683732?check_suite_focus=true#step:8:37132

 however:

1) I can't find the reason, there is a warning instead, 
2) the warning is probably correct, but is it new? doesn't seem connected with the changes I am doing.

**Comment by ye-luo** [comment]
> It seems that some intel test is failing, https://github.com/QMCPACK/qmcpack/runs/8047683732?check_suite_focus=true#step:8:37132
> 
> however:
> 
>     1. I can't find the reason, there is a warning instead, any idea why the test if failing?
> 
>     2. the warning is probably correct, but is it new? doesn't seem connected with the changes I am doing.

I guess the actual error gets clouded by warnings. Have you tried Intel manually?

**Comment by ye-luo** [comment]
```
[ 45%] Building CXX object src/AFQMC/CMakeFiles/afqmc.dir/Drivers/DriverFactory.cpp.o
/scratch/opt/cleanup/qmcpack/src/AFQMC/SlaterDeterminantOperations/mixed_density_matrix.hpp(1014): error: more than one instance of overloaded function "qmcplusplus::afqmc::SlaterDeterminantOperations::batched::generic_sizes" matches the argument list:
            function template "auto qmcplusplus::afqmc::SlaterDeterminantOperations::batched::generic_sizes(const T &)->std::array<std::size_t={unsigned long}, 2UL>"
            function template "auto qmcplusplus::afqmc::SlaterDeterminantOperations::batched::generic_sizes(const T &)->decltype((<expression>))"
            argument types are: (qmcplusplus::afqmc::THCOps::mpi3SPCMatrix)
    int NMO    = (herm ? std::get<1>(generic_sizes(*hermA[0])) : std::get<0>(generic_sizes(*hermA[0])));
                                     ^
          detected during:
            instantiation of "void qmcplusplus::afqmc::SlaterDeterminantOperations::batched::MixedDensityMatrix(std::vector<MatA, std::allocator<MatA>> &, std::vector<MatB, std::allocator<MatB>> &, MatC &&, Tp, TVec &&, Mat &&, Mat &&, IBuffer &, bool, bool) [with MatA=qmcplusplus::afqmc::THCOps::mpi3SPCMatrix *, MatB=boost::multi::array_ptr<qmcplusplus::afqmc::SPComplexType, 2L, qmcplusplus::afqmc::SPComplexType *>, MatC=boost::multi::basic_array<qmcplusplus::afqmc::SPComplexType, 3L,
                      qmcplusplus::afqmc::SPComplexType *, boost::multi::layout_t<3L, boost::multi::size_type={std::make_signed_t<std::size_t={unsigned long}>={MPI_Aint={ptrdiff_t={long}}}}>>, Mat=boost::multi::static_array<qmcplusplus::afqmc::SPComplexType, 3L, qmcplusplus::afqmc::Real3IndexFactorization_batched_v2::device_alloc_type> &, TVec=boost::multi::basic_array<qmcplusplus::afqmc::SPComplexType, 1L, qmcplusplus::afqmc::SPComplexType *, boost::multi::layout_t<1L,
                      boost::multi::size_type={std::make_signed_t<std::size_t={unsigned long}>={MPI_Aint={ptrdiff_t={long}}}}>>, IBuffer=boost::multi::static_array<int, 1L, qmcplusplus::afqmc::HostBufferManager::allocator_t<int>>, Tp=qmcplusplus::afqmc::ComplexType]" at line 251 of "/scratch/opt/cleanup/qmcpack/src/AFQMC/SlaterDeterminantOperations/SlaterDetOperations_serial.hpp"
            instantiation of "void qmcplusplus::afqmc::SlaterDetOperations_serial<Type, BufferManager>::BatchedMixedDensityMatrix(std::vector<MatA, std::allocator<MatA>> &, std::vector<MatB, std::allocator<MatB>> &, MatC &&, qmcplusplus::afqmc::SlaterDetOperations_serial<Type, BufferManager>::T, TVec &&, bool, bool) [with Type=qmcplusplus::afqmc::ComplexType, BufferManager=qmcplusplus::afqmc::DeviceBufferManager, MatA=qmcplusplus::afqmc::THCOps::mpi3SPCMatrix *,
                      MatB=boost::multi::array_ptr<qmcplusplus::afqmc::SPComplexType, 2L, qmcplusplus::afqmc::SPComplexType *>, MatC=boost::multi::basic_array<qmcplusplus::afqmc::SPComplexType, 3L, qmcplusplus::afqmc::SPComplexType *, boost::multi::layout_t<3L, boost::multi::size_type={std::make_signed_t<std::size_t={unsigned long}>={MPI_Aint={ptrdiff_t={long}}}}>>, TVec=boost::multi::basic_array<qmcplusplus::afqmc::SPComplexType, 1L, qmcplusplus::afqmc::SPComplexType *,
                      boost::multi::layout_t<1L, boost::multi::size_type={std::make_signed_t<std::size_t={unsigned long}>={MPI_Aint={ptrdiff_t={long}}}}>>]" at line 297 of "/scratch/opt/cleanup/qmcpack/src/AFQMC/Wavefunctions/Wavefunction.hpp"
            instantiation of "void qmcplusplus::afqmc::Wavefunction::Energy(Args &&...) [with Args=<qmcplusplus::afqmc::WalkerSet &, qmcplusplus::afqmc::ComplexMatrix<qmcplusplus::afqmc::device_allocator<qmcplusplus::afqmc::ComplexType>> &, qmcplusplus::afqmc::ComplexVector<qmcplusplus::afqmc::Real3IndexFactorization_batched_v2::Allocator> &>]" at line 69 of "/scratch/opt/cleanup/qmcpack/src/AFQMC/Estimators/EnergyEstimator.h"

/scratch/opt/cleanup/qmcpack/src/AFQMC/SlaterDeterminantOperations/mixed_density_matrix.hpp(1014): error: no instance of overloaded function "std::get" matches the argument list
            argument types are: (<error-type>)
    int NMO    = (herm ? std::get<1>(generic_sizes(*hermA[0])) : std::get<0>(generic_sizes(*hermA[0])));
```

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**